README.md
[![Build Status](https://travis-ci.org/Yelp/hacheck.png)](https://travis-ci.org/Yelp/hacheck)

**hacheck** is a healthcheck-proxying service. It listens on port 3333, speaks HTTP, and has the following API:

    GET /<protocol>/<service_name>/<port>/<query>

This will check the following locations for service state:

 * `/var/spool/hacheck/all`
 * `/var/spool/hacheck/<service_name>`
 * Depending on the value of `<protocol>`:
  * if `http`: `http://localhost:<port>/<query>`
  * if `tcp`: will attempt to connect to port `<port>` on localhost. `<query>` is currently ignored
  * if `spool`: will only check the spool state
  * if `mysql` and the `mysql_username` and `mysql_password` are set, will do a login and quit on the requested mysql port; `<query>` is ignored and no logical database is selected.

When it does query the actual service check endpoint, **hacheck** MAY cache the value of that query for some amount of time

**hacheck** also comes with the command-line utilities `haup`, `hadown`, and `hastatus`. These take a service name and manipulate the spool files, allowing you to pre-emptively mark a service as "up" or "down".

### Yelp/hacheck versus uber/hacheck

This software was originally developed by [Uber](https://github.com/uber/hacheck).
This version has several features not present in the upstream version:

* By setting the `allow_remote_spool_changes` flag in the config file, it becomes possible to POST changes to spool
  status remotely.
  This is useful if there is some software, like [Paasta](github.com/Yelp/paasta), making decisions centrally about when
  to kill tasks.
* If the client sends the header `X-Haproxy-Server-State`, then we use the port specified in that header and ignore the
  port in the URL. We also use the host address specified in this header instead of localhost.
  This is useful for configurations where an HAProxy has a backend with many servers running on different ports, such
  as when your tasks are running under [Marathon](github.com/mesosphere/marathon). Or if the backends have different
  IPs such as Pods running on Kubernetes.
* If the client sends `X-Nerve-Check-IP`, then we use the IP specified here to check against (rather than the default
  of localhost). This is again useful for Kubernetes Pods where we want to healthcheck with [Nerve](github.com/airbnb/nerve)
  from the hosts network namespace but each Pod runs with its own IP.
* Spool files(downtimes) can optionally have expirations, and store information about when they were created.
* Spool files(downtimes) can apply to a service on a specific port - useful if you have multiple copies of a service on
  the same box.
* It is possible to specify (via `http_headers_to_copy` in the config file) a set of headers that should be forwarded
  to the service being checked.

### Dependencies

**hacheck** is written in Python and makes extensive use of the [tornado](http://www.tornadoweb.org/en/stable/) asynchronous web framework (specifically, it uses the coroutine stuff in Tornado 3). Unit tests use nose and mock.

It runs on Python 2.6 and above, as well as Python 3.3 and above.

### Use cases

Imagine you want to take down the server `web01` for maintenance. Just SSH to it, then (as root) run `hadown all` and wait however long your HAproxy healthchecking interval is. Do your maintenance, then run `haup all` to put it back in service. So easy!

### Configuration

`hacheck` accepts a `-c` flag which should point to a YAML-formatted configuration file. Some notable properties of this file:
* `cache_time`: The duration for which check results may be cached
* `service_name_header`: If set, the name of a header which will be populated with the service name on HTTP checks
* `log_path`: Either the string `"stdout"`, the string `"stderr"`, or a fully-qualified path to a file to write logs to. Uses a [WatchedFileHandler](http://docs.python.org/2/library/logging.handlers.html#watchedfilehandler) and ought to play nicely with logrotate
* `mysql_username`: username to use when logging into mysql for checks
* `mysql_password`: password to use when logging into mysql for checks
* `rlimit_nofile`: set the NOFILE rlimit. If the string "max", will set the rlimit to the hard rlimit; otherwise, will be interpreted as an integer and set to that value.
* `allow_remote_spool_changes`: whether to allow remote control of spool files.

### Monitoring

`hacheck` exports some useful monitoring stuff at the `/status` endpoint. It also exports a count of requests by source-IP and service name on the `/status/count` endpoint.

If the [mutornadomon](https://github.com/uber/mutornadomon) package is available, `hacheck` will import and use it, exposing standard stats about tornado to localhost at `/mutornadomon`

### License

This work is licensed under the [MIT License](http://opensource.org/licenses/MIT), the contents of which can be found at [LICENSE.txt](LICENSE.txt).


debian/changelog
hacheck (0.17.0-yelp2) xenial; urgency=medium

  * Remove trusty
  * Fix tests

 -- Amjad Mashaal <amashaal@yelp.com>  Wed, 19 Jan 2022 09:07:37 -0800

hacheck (0.17.0-yelp1) xenial; urgency=low

  * Run hacheck as multiple processes

 -- David Morrison <drmorr@yelp.com>  Wed, 13 May 2020 16:01:14 -0700

hacheck (0.16.0-yelp1) xenial; urgency=low

  * Remove lucid build targets
  * Upgrade to python 3.6

 -- Stuart Elston <selston@yelp.com>  Mon, 18 Nov 2019 17:34:41 -0800

hacheck (0.15.0-yelp1) xenial; urgency=low

  * Support HTTP 204 responses from proxied services

 -- Stuart Elston <selston@yelp.com>  Tue, 12 Nov 2019 15:48:34 -0800

hacheck (0.14.0-yelp1) xenial; urgency=low

  * Bump tornado to 4.5.3
  * Make healthcheck timeout and tornado max_clients configurable

 -- Stuart Elston <selston@yelp.com>  Thu, 07 Nov 2019 16:46:47 -0800

hacheck (0.13.0-yelp1) lucid; urgency=low

  * Bump tornado and pyyaml version

 -- Mingqi Zhu <mingqiz@yelp.com>  Thu, 31 Oct 2019 10:57:00 -0700
hacheck (0.12.0-yelp1) lucid; urgency=low

  * Support for IPs for Kubernetes

 -- Qui Nguyen <qui@yelp.com>  Thu, 22 Aug 2019 15:52:59 -0700

hacheck (0.11.0-yelp1) lucid; urgency=low

  * Don't use the service name as part of the cache key.

 -- Stephen Arthur <sarth@yelp.com>  Tue, 07 Aug 2018 14:01:22 -0700

hacheck (0.10.0-yelp1) lucid; urgency=low

  * Add https support

 -- Matthew Bentley <bentley@yelp.com>  Mon, 18 Jul 2016 15:08:59 -0700

hacheck (0.9.0-yelp6) lucid; urgency=low

  * Bump yelp-hacheck version.

 -- Keshav Varma <keshav@yelp.com>  Mon, 11 Jul 2016 16:03:33 -0700

hacheck (0.9.0-yelp5) lucid; urgency=low

  * Bump yelp-hacheck version.

 -- Keshav Varma <keshav@yelp.com>  Mon, 11 Jul 2016 15:45:11 -0700

hacheck (0.9.0-yelp4) lucid; urgency=low

  * Bump yelp-hacheck version.

 -- Keshav Varma <keshav@yelp.com>  Mon, 11 Jul 2016 15:17:04 -0700

hacheck (0.9.0-yelp3) lucid; urgency=low

  * No longer follow redirects, instead return
  * Fix tornado version to hopefully avoid memory leaks

 -- Tomas Doran <tdoran@yelp.com>  Tue, 09 Feb 2016 01:37:27 -0800

hacheck (0.9.0-yelp2) lucid; urgency=low

  * Bump yelp-hacheck version.

 -- Evan Krall <krall@yelp.com>  Tue, 21 Jul 2015 14:48:16 -0800

hacheck (0.9.0-yelp1) lucid; urgency=low

  * Bump yelp-hacheck version.

 -- Evan Krall <krall@yelp.com>  Thu, 16 Jul 2015 17:13:33 -0800

hacheck (0.6.1-yelp2) lucid; urgency=low

  * Bump yelp-hacheck version.

 -- John Billings <billings@yelp.com>  Wed, 21 Jan 2015 13:47:21 -0800

hacheck (0.6.1-yelp1) lucid; urgency=low

  * Switch to yelp-hacheck.

 -- John Billings <billings@yelp.com>  Tue, 20 Jan 2015 13:44:39 -0800

hacheck (0.6.1) lucid; urgency=low

  * Initial release.

 -- John Billings <billings@yelp.com>  Wed, 07 Jan 2015 16:12:14 -0800


hacheck/handlers.py
import collections
import logging
import time

import tornado.ioloop
import tornado.httputil
import tornado.httpclient
import tornado.gen
import tornado.web

from . import cache
from . import checker
from . import config
from . import spool

log = logging.getLogger('hacheck')

StatusResponse = collections.namedtuple('StatusResponse', ['code', 'remote_ip', 'ts'])

if hasattr(collections, 'Counter'):
    Counter = collections.Counter  # fast
else:
    def Counter():
        return collections.defaultdict(lambda: 0)

seen_services = {}
service_count = collections.defaultdict(Counter)
last_statuses = {}


def _reset_stats():
    seen_services.clear()
    service_count.clear()
    last_statuses.clear()


class StatusHandler(tornado.web.RequestHandler):
    def get(self):
        stats = {}
        stats['cache'] = cache.get_stats()
        stats['uptime'] = time.time() - self.settings['start_time']

        http_client = tornado.httpclient.AsyncHTTPClient(
            io_loop=tornado.ioloop.IOLoop.current(),
            max_clients=config.config['max_clients'],
        )
        stats['outbound_request_queue_size'] = len(http_client.queue)

        self.set_status(200)
        self.write(stats)


class ListRecentHandler(tornado.web.RequestHandler):
    def get(self):
        now = time.time()
        recency_threshold = int(self.get_argument('threshold', 10 * 60))
        response = []
        for service_name, t in seen_services.items():
            if now - t > recency_threshold:
                continue
            last_status = last_statuses.get(service_name, None)
            if last_status is not None:
                last_status = last_status._asdict()
            response.append((service_name, last_status))
        self.write({
            'seen_services': list(sorted(response)),
            'threshold_seconds': recency_threshold
        })


class ServiceCountHandler(tornado.web.RequestHandler):
    def get(self):
        self.write({'service_access_counts': dict(service_count)})


class BaseServiceHandler(tornado.web.RequestHandler):
    CHECKERS = []

    def maybe_get_host_port_from_haproxy_server_state(self):
        """
        Look for the 'X-Haproxy-Server-State' header and try to parse out the
        'port' value.

        Note that only very recent versions of HAProxy support sending
        the port in the send-state header. In particular you need
        commit 514061c414080701cb046171041a2d00859660e8 from haproxy dev.

        Example server state header:

            X-Haproxy-Server-State: UP 2/3; address=srv2; port=1234;
              name=bck/srv2; node=lb1; weight=1/2; scur=13/22; qcur=0

        returns: the string-typed port if found, else None.
        """

        server_state = self.request.headers.get('X-Haproxy-Server-State', '')

        parts = server_state.split(';')
        parts = [part.strip() for part in parts]
        parts = [part.split('=') for part in parts]
        parts = [part for part in parts if len(part) == 2]

        state = dict(parts)
        return state.get('port'), state.get('address')

    def maybe_get_host_from_nerve_header(self):
        """
        Look for header X-Nerve-Check-IP and use this as the
        host to check rather than localhost. Useful for k8s pods
        which have their own IP and are not reachable on 127.0.0.1
        """

        return self.request.headers.get('X-Nerve-Check-IP', None)

    @tornado.web.asynchronous
    @tornado.gen.coroutine
    def get(self, service_name, port, query):
        header_port, header_host = self.maybe_get_host_port_from_haproxy_server_state()
        if not header_host:
            header_host = self.maybe_get_host_from_nerve_header()
        if header_host and header_host != self.settings['host_ip']:
            host = header_host
        else:
            host = '127.0.0.1'
        port = header_port if header_port else port

        seen_services[service_name] = time.time()
        service_count[service_name][self.request.remote_ip] += 1
        with cache.maybe_bust(self.request.headers.get('Pragma', '') == 'no-cache'):
            port = int(port)
            last_message = ""
            querystr = self.request.query
            for this_checker in self.CHECKERS:
                code, message = yield this_checker(
                    service_name,
                    port,
                    query,
                    host,
                    io_loop=tornado.ioloop.IOLoop.current(),
                    query_params=querystr,
                    headers=self.request.headers,
                )
                last_message = message
                if code > 200:
                    last_statuses[service_name] = StatusResponse(code, self.request.remote_ip, time.time())
                    if code in tornado.httputil.responses:
                        self.set_status(code)
                    else:
                        self.set_status(503)

                    # tornado will complain if we try to write anything (even
                    # an empty string) with an HTTP 204 (No Content)
                    if len(message) > 0:
                        self.write(message)

                    self.finish()
                    break
            else:
                last_statuses[service_name] = StatusResponse(200, self.request.remote_ip, time.time())
                self.set_status(200)
                self.write(last_message)
                self.finish()


class SpoolServiceHandler(BaseServiceHandler):
    CHECKERS = [checker.check_spool]

    def post(self, service_name, port, query):
        if not config.config['allow_remote_spool_changes']:
            self.set_status(403)
            self.write('remote spool changes are not enabled')
            return

        port = int(port) or None
        status = self.get_argument('status')

        if status == 'up':
            spool.up(service_name, port=port)
        elif status == 'down':
            expiration = self.get_argument('expiration', None)
            if expiration is not None:
                expiration = float(expiration)
            reason = self.get_argument('reason')
            creation = self.get_argument('creation', None)
            if creation is not None:
                creation = float(creation)
            spool.down(service_name, reason=reason, port=port, expiration=expiration, creation=creation)
        else:
            self.set_status(400)
            self.write("status must be up or down")
            return

        self.set_status(200)
        self.write("")


class HTTPServiceHandler(BaseServiceHandler):
    CHECKERS = [checker.check_spool, checker.check_http]


class HTTPSServiceHandler(BaseServiceHandler):
    CHECKERS = [checker.check_spool, checker.check_https]


class TCPServiceHandler(BaseServiceHandler):
    CHECKERS = [checker.check_spool, checker.check_tcp]


class MySQLServiceHandler(BaseServiceHandler):
    CHECKERS = [checker.check_spool, checker.check_mysql]


class SMTPServiceHandler(BaseServiceHandler):
    CHECKERS = [checker.check_spool, checker.check_smtp]
